src/lib/parts.ts

```diff
@@ -58,12 +85,33 @@ export class AttributeCommitter {
 
   protected _getValue(): any {
     const strings = this.strings;
+    const parts = this.parts;
     const l = strings.length - 1;
+
+    // If we're assigning an attribute via syntax like:
+    //    attr="${foo}"  or  attr=${foo}
+    // but not
+    //    attr="${foo} ${bar}" or attr="${foo} baz"
+    // then we don't want to coerce the attribute value into one long
+    // string. Instead we want to just return the value itself directly,
+    // so that sanitizeDOMValue can get the actual value rather than
+    // String(value)
+    if (l === 1 && strings[0] === '' && strings[1] === '' && parts[0]) {
+      const v = parts[0].value;
+      if (Array.isArray(v)) {
+        if (v.length === 1) {
+          return v[0];
+        }
+      } else {
+        return v;
+      }
+    }
+
     let text = '';
 
     for (let i = 0; i < l; i++) {
```

Comment:
Maybe add a TODO, and maybe we should loop in Mike, but I feel like we could ideally do some kind of taint-tracking here. If all the values are safe, then I think the result can be safe. Since we don't actually have safe-types to check yet, maybe we can accomplish this via the sanitizer API? Or... maybe that's too much for now.

Reply:
What I'm thinking is that the sanitizer gets both an array of strings and values and can do things like return them untransformed if all values are safe types, validate that the joined string wouldn't start with `'javascript:'`, etc.

Reply:
A safe value doesn't necessarily mean from a trusted source, it could be a value from an untrusted source that has been validated as being safe for a given use case. It's not obvious to me that concatenating safe values produces a safe result.

I think instead what you'd do there is you'd concatenate your inputs first, and then call SafeUrl.sanitize() to get your SafeUrl, which would check that the combined input doesn't start with `javascript:` etc. Then you'd pass your SafeUrl to render. 

The different safe types have different options for construction, e.g. TrustedResourceUrl has a few ways to compose and create a trusted url, but it's even more constrained than SafeUrl because it has higher security requirements.

Reply:
> it could be a value from an untrusted source that has been validated as being safe for a given use case

I thought the point of contextual auto-escaping is that you can't determine if something is safe outside of its context? In this case the values shouldn't be marked as safe independently of their use site. The sanitizer should be able to check that we're not constructing the attribute value `"javascript:doBadThing()"` from  `java${userValue}`, but it shouldn't do the same check on `java${safeValue}` right?

Reply:
It should absolutely do the same check on `java${safeValue}` because the safeValue hasn't been constructed to be safe in any possible configuration. A safe value isn't one that's from a trustworthy source, it's one that was safe at the time it was constructed.

As another example, consider [TrustedResourceUrl](https://google.github.io/closure-library/api/goog.html.TrustedResourceUrl.html), which is a type for a URL that may execute code. Its constructors ensure that it is only constructed in limited ways (e.g. the domain must come from a constant string in source code, with the option of structured parameterization of the path, URL params, etc)
j
As a result, this is safe: `<script src=${trustredResourceUrl}>` but `<script src="http://evil.example.com/${trustedResourceUrl}>` is not safe, and must be blocked.